### PR TITLE
Fix checkpoint values when saving and resetting the tuner state

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

#8578 requires using the checkpoint connector to manage the trainer state before and after tuning.

Part of #8578

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @awaelchli @ananthsub @ninginthecloud @rohitgr7 @akihironitta @borda @carmocca @justusschock

### CODE REVIEW

File: tests/tuner/test_lr_finder.py [hunk 1]
```diff
@@ -343,6 +343,7 @@ def training_step_end(self, outputs):
     trainer.tuner.lr_find(model=model, num_training=num_training)
 
 
+@pytest.mark.xfail
```
Comment (carmocca) [inline tests/tuner/test_lr_finder.py:346]: I couldn't manage to get this test working with just these changes. The test will work with #8578 though

File: pytorch_lightning/tuner/batch_size_scaling.py [hunk 2]
```diff
@@ -59,15 +56,15 @@ def scale_batch_size(
             " Please disable the feature or incorporate the dataloader into the model."
         )
 
-    # Arguments we adjust during the batch size finder, save for restoring
-    __scale_batch_dump_params(trainer)
+    trainer.fit_loop.current_epoch -= 1
+    trainer.fit_loop.global_step -= 1
+    state_dict = deepcopy(trainer.checkpoint_connector.dump_checkpoint())
+    trainer.fit_loop.current_epoch += 1
+    trainer.fit_loop.global_step += 1
```
Comment (carmocca) [inline pytorch_lightning/tuner/batch_size_scaling.py:63]: This extreme ugliness will disappear with #8578

File: pytorch_lightning/tuner/batch_size_scaling.py [hunk 3]
```diff
@@ -59,15 +56,15 @@ def scale_batch_size(
             " Please disable the feature or incorporate the dataloader into the model."
         )
 
-    # Arguments we adjust during the batch size finder, save for restoring
-    __scale_batch_dump_params(trainer)
+    trainer.fit_loop.current_epoch -= 1
+    trainer.fit_loop.global_step -= 1
+    state_dict = deepcopy(trainer.checkpoint_connector.dump_checkpoint())
```
Comment (rohitgr7) [inline pytorch_lightning/tuner/batch_size_scaling.py:61]: this might lead to memory issue?

File: pytorch_lightning/tuner/batch_size_scaling.py [hunk 4]
```diff
@@ -85,59 +86,44 @@ def scale_batch_size(
...
     trainer.callbacks = []  # not needed before full run
     trainer.limit_train_batches = 1.0
-    trainer.optimizers, trainer.strategy.lr_schedulers = [], []  # required for saving
-    trainer.model = model  # required for saving
-
-
-def __scale_batch_restore_params(trainer: "pl.Trainer") -> None:
-    trainer.auto_lr_find = trainer.__dumped_params["auto_lr_find"]
-    trainer.fit_loop.current_epoch = trainer.__dumped_params["current_epoch"]
-    trainer.fit_loop.global_step = trainer.__dumped_params["global_step"]
```
Comment (rohitgr7) [inline pytorch_lightning/tuner/batch_size_scaling.py:133]: calling 
```
trainer = Trainer()
trainer.tune()
trainer.fit()
```
won't work correctly?

File: pytorch_lightning/tuner/batch_size_scaling.py [hunk 5]
```diff
@@ -85,59 +86,44 @@ def scale_batch_size(
...
+def __scale_batch_dump_params(trainer: "pl.Trainer") -> Dict[str, Any]:
+    return {
+        "max_steps": trainer.fit_loop.max_steps,
         "logger": trainer.logger,
         "callbacks": trainer.callbacks,
-        "checkpoint_callback": trainer.checkpoint_callback,
         "auto_scale_batch_size": trainer.auto_scale_batch_size,
+        "auto_lr_find": trainer.auto_lr_find,
         "limit_train_batches": trainer.limit_train_batches,
-        "model": trainer.model,
```
Comment (rohitgr7) [inline pytorch_lightning/tuner/batch_size_scaling.py:114]: might need to update a test here: https://github.com/PyTorchLightning/pytorch-lightning/blob/f7f835fa0e22dc34631aa43d54bf73304c9bda22/tests/tuner/test_scale_batch_size.py#L98

same for lr_finder.

Comment (carmocca) [reply pytorch_lightning/tuner/batch_size_scaling.py:114]: What do you mean? The test is working, no?

File: pytorch_lightning/tuner/batch_size_scaling.py [hunk 6]
```diff
@@ -59,15 +58,17 @@ def scale_batch_size(
...
 
-    # Arguments we adjust during the batch size finder, save for restoring
-    __scale_batch_dump_params(trainer)
+    # Save initial model, that is loaded after batch size is found
+    ckpt_path = os.path.join(trainer.default_root_dir, f".scale_batch_size_{uuid.uuid4()}.ckpt")
+    trainer.fit_loop.current_epoch -= 1
+    trainer.fit_loop.global_step -= 1
+    trainer.save_checkpoint(ckpt_path)
+    trainer.fit_loop.current_epoch += 1
+    trainer.fit_loop.global_step += 1
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/tuner/batch_size_scaling.py:67]: save_checkpoint calls dump_checkpoint which itself saves an increased current_epoch and global step count.

So either we decrease it before calling this or we save the checkpoint and then load it again and change the values

[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/tuner/batch_size_scaling.py:67]: Note that these decrements and increments will disappear in a follow-up: #8578

File: pytorch_lightning/tuner/batch_size_scaling.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/tuner/batch_size_scaling.py:133]: Why not?

File: pytorch_lightning/tuner/batch_size_scaling.py (hunk 5 above)
Comment (rohitgr7) [reply pytorch_lightning/tuner/batch_size_scaling.py:114]: yes.. I mean since a few of the arguments are removed here, maybe need to remove them from the test too?

File: pytorch_lightning/tuner/batch_size_scaling.py (hunk 4 above)
Comment (rohitgr7) [reply pytorch_lightning/tuner/batch_size_scaling.py:133]: oh okay. nvm. I see they will be reloaded from the checkpoint itself.

File: pytorch_lightning/tuner/lr_finder.py [hunk 7]
```diff
@@ -270,27 +254,33 @@ def lr_find(
...
+    # avoid lr find being called multiple times
+    trainer.auto_lr_find = False
+    # Use special lr logger callback
+    trainer.callbacks = [_LRCallback(num_training, early_stop_threshold, progress_bar_refresh_rate=1)]
+    # No logging
+    trainer.logger = DummyLogger() if trainer.logger is not None else None
+    # Max step set to number of iterations
+    trainer.fit_loop.max_steps = num_training
+    # Required for saving the model
+    trainer.optimizers, trainer.strategy.lr_schedulers = [], []
```
Comment (rohitgr7) [inline pytorch_lightning/tuner/lr_finder.py:276]: last ques: won't this be an empty list already?

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/tuner/lr_finder.py (hunk 7 above)
Comment (carmocca) [reply pytorch_lightning/tuner/lr_finder.py:276]: I'm also not sure about why this is needed, I just copied it verbatim from where it was before

File: pytorch_lightning/tuner/batch_size_scaling.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/tuner/batch_size_scaling.py:114]: This test was helpful to validate my implementation (had some bugs before) so I prefer keeping those

File: pytorch_lightning/tuner/lr_finder.py (hunk 7 above)
Comment (rohitgr7) [reply pytorch_lightning/tuner/lr_finder.py:276]: it was maybe done because a temp checkpoint was being created after this, so to avoid saving the optimizer/scheduler state_dict. But now checkpoint is created before the reset, so this won't have any effect.

Comment (carmocca) [reply pytorch_lightning/tuner/lr_finder.py]: Removed :)

**Review by ethanwharris — APPROVED** [review]

File: pytorch_lightning/tuner/batch_size_scaling.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/tuner/batch_size_scaling.py:133]: @carmocca

**Review by akihironitta — APPROVED** [review]

File: pytorch_lightning/tuner/lr_finder.py [hunk 8]
```diff
@@ -270,27 +254,31 @@ def lr_find(
...
 
-def __lr_finder_restore_params(trainer, model):
-    trainer.auto_lr_find = trainer.__dumped_params["auto_lr_find"]
-    trainer.logger = trainer.__dumped_params["logger"]
-    trainer.callbacks = trainer.__dumped_params["callbacks"]
-    trainer.fit_loop.global_step = trainer.__dumped_params["global_step"]
-    trainer.fit_loop.max_steps = trainer.__dumped_params["max_steps"]
-    trainer.fit_loop.current_epoch = trainer.__dumped_params["current_epoch"]
-    del trainer.__dumped_params
+def __lr_finder_reset_params(trainer: "pl.Trainer", num_training: int, early_stop_threshold: float) -> None:
```
Comment (akihironitta) [inline pytorch_lightning/tuner/lr_finder.py:266]: Completely unrelated to this PR itself, but just out of my curiosity, is there any convention on usage of double underscores (`__`) and single underscore (`_`) in the project?

Comment (carmocca) [reply pytorch_lightning/tuner/lr_finder.py:266]: I personally dislike double underscore (private) because we rely a lot in inheritance where it can become a problem, also because we rarely have naming conflicts that would motivate it

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 9]
```diff
@@ -2074,7 +2074,7 @@ def model(self, model: torch.nn.Module) -> None:
             model: The LightningModule, possibly wrapped into DataParallel or DistributedDataParallel, depending
                 on the backend.
         """
-        self.strategy.model = model
+        self.strategy.connect(model)
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:2077]: I believe I can undo this change, this would only impact TPUs

```suggestion
        self.strategy.model = model
```

But I wonder if calling connect is the best solution long-term as that's where Strategy authors can add their logic.
